examples/dataflow-csvio/src/main/java/com/google/example/csvio/SortContextualHeadersAndRows.java

```diff
@@ -0,0 +1,427 @@
+      PCollectionView<Map<String, Long>> mappedHeaderPositionView =
+          resourceIdHeaderPosition.apply(View.asMap());
+      PCollectionTuple pct =
+          input.apply(
+              TAG_BASE + "/" + SortMappedFixedPositionHeadersAndRowsFn.class.getSimpleName(),
+              ParDo.of(new SortMappedFixedPositionHeadersAndRowsFn(mappedHeaderPositionView))
+                  .withSideInput("", mappedHeaderPositionView)
+                  .withOutputTags(HEADERS, TupleTagList.of(ROWS)));
+
+      return new Result(pct);
+    }
+  }
+
+  /**
+   * A {@link PTransform} responsible for bundling the raw {@link ContextualTextIO#read()} {@link
+   * Row} {@link PCollection} output into the corresponding CSV headers and rows in the setting
+   * where the header is expected at the first non-empty line of the CSV file.
+   */
+  static class SortMinimumHeaderAndRows extends PTransform<PCollection<Row>, Result> {
+
+    private static final String TAG_BASE =
+        SortContextualHeadersAndRows.TAG_BASE
+            + "/"
+            + SortMinimumHeaderAndRows.class.getSimpleName();
+
+    @Override
+    public Result expand(PCollection<Row> input) {
+      PCollection<KV<String, Long>> resourceIdRecordOffsetKV =
+          input
+              .apply(
+                  TAG_BASE + "/" + IsNotEmptyFn.class.getSimpleName(),
+                  Filter.by(new IsNotEmptyFn()))
+              .apply(
+                  TAG_BASE + "/MapElements/" + RowKVFn.class.getSimpleName(),
+                  MapElements.via(new RowKVFn()));
+
+      PCollection<KV<String, Long>> minimumResourceIdRecordOffsetKV =
+          resourceIdRecordOffsetKV.apply(TAG_BASE + "/Min", Min.longsPerKey());
+
+      PCollectionView<Map<String, Long>> minimumResourceIdHeaderMapView =
+          minimumResourceIdRecordOffsetKV.apply(TAG_BASE + "/Min/View", View.asMap());
+
+      PCollectionTuple pct =
+          input.apply(
+              TAG_BASE + "/" + SortMappedFixedPositionHeadersAndRowsFn.class.getSimpleName(),
+              ParDo.of(new SortMappedFixedPositionHeadersAndRowsFn(minimumResourceIdHeaderMapView))
+                  .withSideInput("", minimumResourceIdHeaderMapView)
+                  .withOutputTags(HEADERS, TupleTagList.of(ROWS)));
+
+      return new Result(pct);
+    }
+  }
+
+  /**
+   * The {@link DoFn} responsible for sorting a fixed header position with the subsequent CSV
+   * records.
+   */
+  static class SortFixedPositionHeadersAndRowsFn extends DoFn<Row, Row> {
+
+    private final PCollectionView<Long> headerPositionView;
+
+    SortFixedPositionHeadersAndRowsFn(PCollectionView<Long> headerPositionView) {
+      this.headerPositionView = headerPositionView;
+    }
+
+    @ProcessElement
+    public void process(ProcessContext ctx) {
+      Long headerPosition = ctx.sideInput(this.headerPositionView);
+      Row input = ctx.element();
+      RecordWithMetadataHelper record = RecordWithMetadataHelper.of(input);
```

Comment:
Since calls to RecordWithMetadataHelper happens per element this leads to creation & destruction of several objects causing excess GC overload.  Better to move RecordWithMetadataHelper as standalone class (in its own file) with static methods only and can be called passing row as Input